lib/pbench/cli/agent/commands/results/move.py

```diff
@@ -0,0 +1,257 @@
+                        )
+                        failures += 1
+                        # If we can't hold up the contract of marking a
+                        # directory as copied remotely, we exit the loop that
+                        # is processing result directories.  If we can't
+                        # create an empty file on the file system where the
+                        # result directory lives, it likely indicates bigger
+                        # problems.
+                        break
+                else:
+                    try:
+                        shutil.rmtree(result_dir)
+                    except OSError:
+                        self.logger.error(
+                            "Failed to remove the %s directory hierarchy", result_dir
+                        )
+                        failures += 1
+                        # If we can't hold up the contract of removing the
+                        # local directory tree that was copied, we exit the
+                        # loop that is processing result directories.  Not
+                        # being able to remove the local directory tree will
+                        # usually indicate a serious problem that needs to be
+                        # resolved before doing anything else.
+                        break
+
+        action = "copied" if copy_only else "moved"
+        click.echo(
+            f"Status: total # of result directories considered {no_of_tb:d},"
+            f" successfully {action} {runs_copied:d}, encountered"
+            f" {failures:d} failures"
+        )
+
+        return 0 if failures == 0 else 1
+
+
+@click.command()
+@common_options
+@click.option(
+    "--controller",
+    required=False,
+    envvar="PBENCH_CONTROLLER",
+    default="",
+    prompt=False,
+    help="Override the default controller name",
+)
+@click.option(
+    "--token",
+    required=False,
+    envvar="PBENCH_ACCESS_TOKEN",
+    prompt=False,
+    help="pbench server authentication token",
+)
+@click.option(
+    "--copy",
+    is_flag=True,
+    help="Only perform the remote copy, do not remove local data",
+)
+@click.option(
+    "--xz-single-threaded",
+    is_flag=True,
+    help="Use single threaded compression with 'xz'",
+)
+@click.option(
+    "--show-server",
+    help="Display information about the pbench server where the result(s) will be moved (Not implemented)",
+)
+@pass_cli_context
+def main(
+    context: CliContext,
+    controller: str,
+    token: str,
+    copy: bool,
+    xz_single_threaded: bool,
+    show_server: str,
+):
+    clk_ctx = click.get_current_context()
+
+    if controller:
+        context.controller = controller
+    else:
+        controller = os.environ.get("_pbench_full_hostname", socket.getfqdn())
+        if not controller:
+            click.echo(
+                "INTERNAL ERROR - unable to determine the controller name, could not fetch the FQDN of the host",
+                err=True,
+            )
```

Comment:
We don't require the `--controller` switch, do we?  So wouldn't emitting a message telling the user to provide a value be against our interface goals?  If our fallback of calling `socket.getfqdn()` does not provide a FQDN name for the host, we don't have a fallback to use, which is not the fault of the user, right?  So why not inform the user of a work-around, as you suggest, while still letting them know a base assumption is broken?

Reply:
To be honest, I get real fuzzy real fast on who is responsible for which domains with this stuff.

From what I've seen, it seems clear that we currently need the controller to be set correctly, otherwise we end up with problems trying to unpack the tarball (which, I have to admit, sort of surprises me, but I assume that's just my limited knowledge).  We have three spots to try to get the controller from -- the command line, the environment variable, and the current host -- I don't know who is responsible for each of them.  The command line seems like it would be the user, but if we're invoking this command _on behalf of the user_, then it's on us (although, I don't know where _we_ would get the value from).  The environment variable could be something that we set up, or it could be there for the user's convenience; I have no idea.  But, the configuration of the host is _not_ in our wheelhouse (right??), so, if the fallback doesn't work, that's not a short-coming our our product -- hence, _not_ an "internal error" -- that problem is something the user has to fix (I would think...) if they want us to get the controller name that way.

So, I'm all for telling the user what happened, and I'm in favor of explaining how to fix it, but labeling something an internal error when it's ultimately the result of something the user did (or didn't do), doesn't help the user, and it won't be good for us.

Reply:
The user is not responsible for the host not having an FQDN, is it?

Reply:
Well, _somebody_ configured the host that this stuff is running on.  I'm gonna go out on a limb here and say that, if it wasn't the user himself, at least he's got a closer relationship to the sysadmin than our code does.

We don't consider file system access problems to be internal errors (right?), so why should problems accessing other parts of the system configuration be deemed internal errors?

Reply:
Arguably we should report them as INTERNAL ERRORS because they are not errors the user caused, but are fundamental broken assumptions.

It is okay to disagree on that point.

If just removing `INTERNAL ERROR` from the message is what you are looking for, then I'll remove it.

Reply:
I would suggest that an internal error is the result of software detecting of an incorrect or inconsistent piece of state that it maintains.  However, the FSDN is an _external_ piece of state -- it's not ours to set or maintain.

We can quibble about whether the user caused it or not.  I'm not sure who "the user" is in this context...mostly it seems to me to be people who provision machines, install stuff on them, and then run Pbench...in which case, it would be up to them to make sure that the system has a FQDN configured.  In any case, it's the user (not Pbench or its maintainers) who has any realistic hope of getting the system configuration corrected if it's missing.

What I'm looking for is a good experience for the user, and, secondarily, for us.  If we flag an error in the user's configuration as an "internal Pbench error", then the user is unlikely to pursue fixing it themself and is instead (best case...) going to come after us to fix it.  (In the worst case, they will label our software as "buggy" and go elsewhere.)

I don't know how we expect to come up with the controller name, but, if we fail, and the user can address the problem, then I think we should empower and encourage the user to dig us out of the collective hole that we're in, and removing `INTERNAL ERROR` would be an excellent first step.